src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -391,12 +390,15 @@ private void ResetAndAbort(ConnectionAbortedException abortReason, Http2ErrorCod
             AbortCore(abortReason);
         }
 
-        private void AbortCore(ConnectionAbortedException abortReason)
+        private void AbortCore(Exception abortReason)
         {
-            base.Abort(abortReason);
+            // Call OnIOCompleted() which closes the output prior to poisoning the request body stream or pipe to
+            // ensure that an app that completes early due to the abort doesn't result in header frames being sent.
+            OnInputOrOutputCompleted();
 
             // Unblock the request body.
-            RequestBodyPipe.Writer.Complete(new IOException(CoreStrings.Http2StreamAborted, abortReason));
+            PoisonRequestBodyStream(abortReason);
```

Comment:
Is there a risk to calling this multiple times?

Reply:
It could wind up changing the exception type thrown from the request body stream. AbortCore should only ever be called once per stream though.